File: JavaScript/maxProfit.js
/*
Say you have an array prices for which the ith element is the price of a given stock on day i.

Design an algorithm to find the maximum profit. You may complete as many transactions as you like (i.e., buy one and sell one share of the stock multiple times).

Note: You may not engage in multiple transactions at the same time (i.e., you must sell the stock before you buy again).

*/

let maxProfit = function(prices) {
	let holding = false;
	let profit = 0;
	let boughtPrice = 0;
	for (let i = 0; i < prices.length; i++) {
		let tomorrowsPrice = prices[i + 1];
		let todaysPrice = prices[i];
		if (holding == true) {
			if (tomorrowsPrice < todaysPrice) {
				profit = profit + todaysPrice - boughtPrice;
				holding = false;
			}
		} else if (holding == false) {
			if (tomorrowsPrice > todaysPrice) {
				boughtPrice = todaysPrice;
				holding = true;
			}
		}
	}

	if (holding == true) {
		profit = profit + prices[prices.length - 1] - boughtPrice;
	}

	return profit;
};

prices = [
	1,
	2,
	3,
	4,
	5,
];
answer = maxProfit(prices);

console.log(answer);

File: python/maxProfit.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if len(prices) < 2:
            return 0

        i = 0
        j = i + 1
        profit = 0
        while (i < len(prices) - 1):
            if (prices[i] < prices[j]):
                # buy at i
                while (j < len(prices) - 1 and prices[j] <= prices[j+1]):
                    j += 1

                profit += (prices[j] - prices[i])
                i = j + 1
                j = i + 1
            else:
                i += 1
                j += 1
        return profit


def main():
    print(Solution().maxProfit([7, 1, 5, 3, 6, 4]), 'expecting 7')
    print(Solution().maxProfit([1, 2, 3, 4, 5]), 'expecting 4')
    print(Solution().maxProfit([7, 6, 4, 3, 1]), 'expecting 0')
    print(Solution().maxProfit([1, 2]), 'expecting 1')
    print(Solution().maxProfit([1, 1]), 'expecting 0')


if __name__ == '__main__':
    main()
